src/types/impl-trait.md

```diff
@@ -111,10 +123,189 @@ fn foo() -> impl Trait {
+This is similar to how all code paths in a function with a return-position-impl-trait must return the same type:
+
+```rust,ignore
+fn foo() -> impl Debug {
+    if true {
+        return 42;
+    }
+    "42" // ERROR: another return site returns an `i32`
+}
+
+type Bar = impl Debug;
+fn bar() -> Bar {
+    42
+}
+fn boo() -> Bar {
+    "42" // ERROR: another function uses `i32` for `Bar`
+}
+```
+
+#### Argument position
+
+Without knowing the hidden type, we can still use the opaque type and use its trait bounds. In this case we can use the `Debug` trait to render it:
+
+```rust,ignore
+fn bop(x: Bar) {
+    println!("{x:?}");
+}
+```
+
+Note that this is very different from using `impl Trait` in argument position, as there is no anonymous generic parameter introduced.
+
+Binding a hidden type works in both directions, not just assigning a hidden type value to the opaque type, but also reading an opaque type into a hidden type value:
+
+```rust,ignore
+fn bup(x: Bar) {
+    let x: i32 = x;
+}
+```
+
+This does not "reveal" the hidden type. It binds an explicitly known `i32` type as the hidden type of `Bar` and will error if that's not the hidden type everywhere else, too.
+
+As a last usage, you can avoid binding any hidden types and just use the type-alias-impl-trait by just forwarding it elsewhere:
+
+```rust,ignore
+fn burp(x: Bar) -> Bar {
+    x
+}
+```
+
+#### Binding types
+
+You can also use type-alias-impl-trait for the type
+of local variables, constants, statics, ...
+
+```rust,ignore
+let x: Bar = 42;
+const X: Bar = 42;
+static Y: Bar = 42;
+```
+
+#### Nested in other types
+
+You can use type-alias-impl-trait in other types:
+
+```rust,ignore
+struct MyStruct {
+    bar: Bar,
+}
+```
+
+and use it just like other uses of `Bar`:
+
+```rust,ignore
+fn foo(my_struct: &MyStruct) {
+    println!("{:?}", my_struct.bar);
+}
+fn new() -> MyStruct {
+    MyStruct {
+        bar: 42
+    }
+}
+```
+
+#### Usage in trait impls
+
+Since type-alias-impl-trait can be referenced anywhere a type alias could be, this also means you can use them in `impl` blocks:
+
+```rust,ignore
+type Foo = impl Trait;
+
+impl Bar for Foo {}
+```
+
+There's a huge caveat though: now it's possible for there to be an impl for an opaque type *and* its hidden type:
+
+```rust,ignore
+type Foo = impl Trait;
+fn foo() -> Foo {}
+
+impl Bar for Foo {}
+impl Bar for () {} // ERROR conflicts with `impl Bar for Foo`
+```
+
+This check is *not* done by revealing the hidden type, but by checking whether a type could be a hidden type for that specific opaque type. So the following program is legal:
+
+```rust
+#![feature(type_alias_impl_trait)]
+trait Trait {}
+impl Trait for () {}
+type Foo = impl Trait;
+fn foo() -> Foo {}
+
+trait Bar {}
+impl Bar for Foo {}
+impl Bar for i32 {}
+```
+
+This is legal, because `i32` could not possibly be a hidden type of `Foo`, because it doesn't implement `Trait` wich is a requirement for all hypothetical hidden types of `Foo`.
```

Comment:
I find this pretty surprising -- it means adding a trait impl is a direct breaking change, not just because of changes to e.g. method call resolution or similar. Is this limited to the same crate perhaps?

Reply:
Yes, this is limited to impls for traits in the same crate.